File: lld/problems/notification-system/java/src/NotificationSystemTest.java
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class NotificationSystemTest {

    public static void main(String[] args) {
        runAll();
    }

    public static void runAll() {
        testRenderTemplate();
        testNotifyDispatchesPerPreference();
        testNotifyUnknownUserThrows();
        testRetrySucceedsAfterNFailures();
        testRetryGivesUpAfterMaxAttempts();
        testNotifyMultipleChannelsIndependentRetry();
        System.out.println("All NotificationSystemTest cases passed.");
    }

    private static void testRenderTemplate() {
        Map<String, String> data = new HashMap<>();
        data.put("name", "Alice");
        assertEquals("Hello Alice", TemplateRenderer.render("Hello {name}", data), "single placeholder");

        data.put("orderId", "42");
        assertEquals("Hello Alice, your order 42 shipped",
                TemplateRenderer.render("Hello {name}, your order {orderId} shipped", data),
                "multiple placeholders");

        Map<String, String> partial = new HashMap<>();
        partial.put("name", "Carl");
        assertEquals("Hi Carl, code {code}", TemplateRenderer.render("Hi {name}, code {code}", partial), "missing key left untouched");

        assertEquals("plain message", TemplateRenderer.render("plain message", new HashMap<>()), "no placeholders");
        assertEquals("Hello {name", TemplateRenderer.render("Hello {name", data), "unterminated brace");
    }

    private static void testNotifyDispatchesPerPreference() {
        EmailChannel email = new EmailChannel();
        SMSChannel sms = new SMSChannel();
        PushChannel push = new PushChannel();

        NotificationService service = new NotificationService(new RetryPolicy(1, 0));
        service.registerChannel(email);
        service.registerChannel(sms);
        service.registerChannel(push);

        service.setPreferences("u1", Channel.EMAIL, Channel.SMS);
        service.setPreferences("u2", Channel.PUSH);

        Map<String, String> data = new HashMap<>();
        data.put("name", "Ann");
        List<SendResult> results = service.notify("u1", "u1@example.com", "Hello {name}", data);

        if (results.size() != 2) {
            throw new AssertionError("expected 2 results for u1, got " + results.size());
        }
        for (SendResult r : results) {
            if (!r.success()) {
                throw new AssertionError("unexpected send error on channel " + r.channel);
            }
        }
        assertEquals(1, email.sent().size(), "email sent count");
        assertEquals("Hello Ann", email.sent().get(0).message, "email rendered message");
        assertEquals(1, sms.sent().size(), "sms sent count");
        assertEquals(0, push.sent().size(), "push should not receive u1's message");

        Map<String, String> data2 = new HashMap<>();
        data2.put("name", "Bo");
        service.notify("u2", "device-token", "Ping {name}", data2);
        assertEquals(1, push.sent().size(), "push sent count for u2");
        assertEquals("Ping Bo", push.sent().get(0).message, "push rendered message");
    }

    private static void testNotifyUnknownUserThrows() {
        NotificationService service = new NotificationService(RetryPolicy.defaultPolicy());
        service.registerChannel(new EmailChannel());

        try {
            service.notify("ghost", "x", "hi", new HashMap<>());
            throw new AssertionError("expected IllegalArgumentException for unknown user");
        } catch (IllegalArgumentException expected) {
            // expected
        }
    }

    private static void testRetrySucceedsAfterNFailures() {
        EmailChannel underlying = new EmailChannel();
        FlakyChannel flaky = new FlakyChannel(underlying, 2);

        NotificationService service = new NotificationService(new RetryPolicy(3, 1));
        service.registerChannel(flaky);
        service.setPreferences("u1", Channel.EMAIL);

        Map<String, String> data = new HashMap<>();
        data.put("orderId", "7");
        List<SendResult> results = service.notify("u1", "u1@example.com", "Order {orderId} shipped", data);

        assertEquals(1, results.size(), "result count");
        SendResult r = results.get(0);
        if (!r.success()) {
            throw new AssertionError("expected eventual success, got err: " + r.error);
        }
        assertEquals(3, r.attempts, "attempts (2 failures + 1 success)");
        assertEquals(3, flaky.attempts(), "channel invocation count");
        assertEquals(1, underlying.sent().size(), "underlying successful sends");
        assertEquals("Order 7 shipped", underlying.sent().get(0).message, "underlying message content");
    }

    private static void testRetryGivesUpAfterMaxAttempts() {
        AlwaysFailChannel failing = new AlwaysFailChannel(Channel.SMS);

        NotificationService service = new NotificationService(new RetryPolicy(4, 1));
        service.registerChannel(failing);
        service.setPreferences("u1", Channel.SMS);

        Map<String, String> data = new HashMap<>();
        data.put("name", "X");
        List<SendResult> results = service.notify("u1", "5551234", "hi {name}", data);

        assertEquals(1, results.size(), "result count");
        SendResult r = results.get(0);
        if (r.success()) {
            throw new AssertionError("expected send to ultimately fail");
        }
        assertEquals(4, r.attempts, "attempts equal to MaxAttempts");
        assertEquals(4, failing.attempts(), "channel invocation count");
    }

    private static void testNotifyMultipleChannelsIndependentRetry() {
        EmailChannel email = new EmailChannel();
        FlakyChannel flakySms = new FlakyChannel(new SMSChannel(), 1);

        NotificationService service = new NotificationService(new RetryPolicy(2, 1));
        service.registerChannel(email);
        service.registerChannel(flakySms);
        service.setPreferences("u1", Channel.EMAIL, Channel.SMS);

        List<SendResult> results = service.notify("u1", "recipient", "msg", new HashMap<>());
        assertEquals(2, results.size(), "result count");
        for (SendResult r : results) {
            if (!r.success()) {
                throw new AssertionError("expected both channels to eventually succeed, channel " + r.channel + " failed: " + r.error);
            }
        }
        assertEquals(1, results.get(0).attempts, "email succeeds on first attempt");
        assertEquals(2, results.get(1).attempts, "sms succeeds on second attempt");
    }

    private static void assertEquals(Object expected, Object actual, String label) {
        if (!expected.equals(actual)) {
            throw new AssertionError(label + ": expected " + expected + " but got " + actual);
        }
    }
}

File: lld/problems/notification-system/go/notification_system_test.go
package notificationsystem

import (
	"testing"
	"time"
)

func TestRenderTemplate(t *testing.T) {
	tests := []struct {
		name     string
		template string
		data     map[string]string
		want     string
	}{
		{
			name:     "single placeholder",
			template: "Hello {name}",
			data:     map[string]string{"name": "Alice"},
			want:     "Hello Alice",
		},
		{
			name:     "multiple placeholders",
			template: "Hello {name}, your order {orderId} shipped",
			data:     map[string]string{"name": "Bob", "orderId": "42"},
			want:     "Hello Bob, your order 42 shipped",
		},
		{
			name:     "missing key left untouched",
			template: "Hi {name}, code {code}",
			data:     map[string]string{"name": "Carl"},
			want:     "Hi Carl, code {code}",
		},
		{
			name:     "no placeholders",
			template: "plain message",
			data:     map[string]string{"unused": "x"},
			want:     "plain message",
		},
		{
			name:     "unterminated brace",
			template: "Hello {name",
			data:     map[string]string{"name": "Dee"},
			want:     "Hello {name",
		},
		{
			name:     "empty template",
			template: "",
			data:     map[string]string{},
			want:     "",
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got := RenderTemplate(tt.template, tt.data)
			if got != tt.want {
				t.Fatalf("RenderTemplate(%q) = %q, want %q", tt.template, got, tt.want)
			}
		})
	}
}

func TestNotifyDispatchesPerPreference(t *testing.T) {
	email := NewEmailChannel()
	sms := NewSMSChannel()
	push := NewPushChannel()

	svc := NewNotificationService(RetryPolicy{MaxAttempts: 1})
	svc.RegisterChannel(email)
	svc.RegisterChannel(sms)
	svc.RegisterChannel(push)

	svc.SetPreferences("u1", Email, SMS)
	svc.SetPreferences("u2", Push)

	results, err := svc.Notify("u1", "u1@example.com", "Hello {name}", map[string]string{"name": "Ann"})
	if err != nil {
		t.Fatalf("unexpected err: %v", err)
	}
	if len(results) != 2 {
		t.Fatalf("expected 2 results for u1, got %d", len(results))
	}
	for _, r := range results {
		if r.Err != nil {
			t.Fatalf("unexpected send error on channel %s: %v", r.Channel, r.Err)
		}
	}
	if len(email.Sent()) != 1 || email.Sent()[0].Message != "Hello Ann" {
		t.Fatalf("expected email to receive rendered message, got %+v", email.Sent())
	}
	if len(sms.Sent()) != 1 {
		t.Fatalf("expected sms to receive one message, got %+v", sms.Sent())
	}
	if len(push.Sent()) != 0 {
		t.Fatalf("expected push to receive nothing for u1, got %+v", push.Sent())
	}

	if _, err := svc.Notify("u2", "device-token", "Ping {name}", map[string]string{"name": "Bo"}); err != nil {
		t.Fatalf("unexpected err: %v", err)
	}
	if len(push.Sent()) != 1 || push.Sent()[0].Message != "Ping Bo" {
		t.Fatalf("expected push to receive rendered message for u2, got %+v", push.Sent())
	}
}

func TestNotifyUnknownUserReturnsError(t *testing.T) {
	svc := NewNotificationService(DefaultRetryPolicy())
	svc.RegisterChannel(NewEmailChannel())

	if _, err := svc.Notify("ghost", "x", "hi", nil); err == nil {
		t.Fatalf("expected error for user with no preferences")
	}
}

func TestRetrySucceedsAfterNFailures(t *testing.T) {
	underlying := NewEmailChannel()
	flaky := NewFlakyChannel(underlying, 2) // fails twice, succeeds on 3rd attempt

	svc := NewNotificationService(RetryPolicy{MaxAttempts: 3, Delay: time.Millisecond})
	svc.RegisterChannel(flaky)
	svc.SetPreferences("u1", Email)

	results, err := svc.Notify("u1", "u1@example.com", "Order {orderId} shipped", map[string]string{"orderId": "7"})
	if err != nil {
		t.Fatalf("unexpected err: %v", err)
	}
	if len(results) != 1 {
		t.Fatalf("expected 1 result, got %d", len(results))
	}
	r := results[0]
	if r.Err != nil {
		t.Fatalf("expected eventual success, got err: %v", r.Err)
	}
	if r.Attempts != 3 {
		t.Fatalf("expected 3 attempts (2 failures + 1 success), got %d", r.Attempts)
	}
	if flaky.Attempts() != 3 {
		t.Fatalf("expected channel to have been called 3 times, got %d", flaky.Attempts())
	}
	if len(underlying.Sent()) != 1 || underlying.Sent()[0].Message != "Order 7 shipped" {
		t.Fatalf("expected underlying channel to record exactly one successful send, got %+v", underlying.Sent())
	}
}

func TestRetryGivesUpAfterMaxAttempts(t *testing.T) {
	failing := NewAlwaysFailChannel(SMS)

	svc := NewNotificationService(RetryPolicy{MaxAttempts: 4, Delay: time.Millisecond})
	svc.RegisterChannel(failing)
	svc.SetPreferences("u1", SMS)

	results, err := svc.Notify("u1", "5551234", "hi {name}", map[string]string{"name": "X"})
	if err != nil {
		t.Fatalf("unexpected err: %v", err)
	}
	if len(results) != 1 {
		t.Fatalf("expected 1 result, got %d", len(results))
	}
	r := results[0]
	if r.Err == nil {
		t.Fatalf("expected send to ultimately fail")
	}
	if r.Attempts != 4 {
		t.Fatalf("expected exactly MaxAttempts=4 attempts, got %d", r.Attempts)
	}
	if failing.Attempts() != 4 {
		t.Fatalf("expected channel to have been called 4 times, got %d", failing.Attempts())
	}
}

func TestNotifyMultipleChannelsIndependentRetry(t *testing.T) {
	email := NewEmailChannel()
	flakySMS := NewFlakyChannel(NewSMSChannel(), 1)

	svc := NewNotificationService(RetryPolicy{MaxAttempts: 2, Delay: time.Millisecond})
	svc.RegisterChannel(email)
	svc.RegisterChannel(flakySMS)
	svc.SetPreferences("u1", Email, SMS)

	results, err := svc.Notify("u1", "recipient", "msg", nil)
	if err != nil {
		t.Fatalf("unexpected err: %v", err)
	}
	if len(results) != 2 {
		t.Fatalf("expected 2 results, got %d", len(results))
	}
	for _, r := range results {
		if r.Err != nil {
			t.Fatalf("expected both channels to eventually succeed, channel %s failed: %v", r.Channel, r.Err)
		}
	}
	if results[0].Attempts != 1 {
		t.Fatalf("expected email to succeed on first attempt, got %d", results[0].Attempts)
	}
	if results[1].Attempts != 2 {
		t.Fatalf("expected sms to succeed on second attempt, got %d", results[1].Attempts)
	}
}
